### PR TITLE
floats: add functions for handling NaN payloads

### PR DESCRIPTION
**Pull request by kortschak**
Addresses https://github.com/gonum/gonum/pull/50#issuecomment-306936995.

Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
We can add sign handling if wanted `NaN(sign int, payload uint64) float64`; sign is sometimes used for things here apparently, though `math.Copysign` will do that for both directions (NaN->payload and payload->NaN).

File: floats/floats.go [hunk 1]
```diff
@@ -490,6 +490,30 @@ func MulTo(dst, s, t []float64) []float64 {
...
+const (
+	nanBits = 0x7ff8000000000000
+	nanMask = 0xfff8000000000000
+)
+
+// NaN returns an IEEE 754 "quiet not-a-number" value with the
+// payload specified in the low 51 bits of payload.
+// The NaN returned by math.NaN has a bit pattern equal to NaN(1).
+func NaN(payload uint64) float64 {
+	payload &^= nanMask
```
Comment (btracey) [inline floats/floats.go]: Is there a reason to have this two lines?
`math.Float64frombits(nanBits | (payload &^ nanMask))`

File: floats/floats.go [hunk 2]
```diff
@@ -490,6 +490,30 @@ func MulTo(dst, s, t []float64) []float64 {
...
+
+// NaNPayload returns the lowest 51 bits of an IEEE 754 "quiet
+// not-a-number" and true, or zero and false if f is not NaN
+// or is not quiet.
+func NaNPayload(f float64) (payload uint64, ok bool) {
+	b := math.Float64bits(f)
+	if b&nanBits != nanBits {
+		return 0, false
+	}
+	return b &^ nanMask, true
```
Comment (btracey) [inline floats/floats.go]: Sorry, is this right? I'm not great at bitwise arithmetic yet. `b &^ nanMask` keeps all of the `nanMask` bits of `b` and turns all of the rest to 0, doesn't it? It seems like that's exactly the opposite of what we want. The test passes, so I must not be understanding something ...

File: floats/floats.go [hunk 3]
```diff
@@ -490,6 +490,30 @@ func MulTo(dst, s, t []float64) []float64 {
...
 
+const (
+	nanBits = 0x7ff8000000000000
+	nanMask = 0xfff8000000000000
+)
+
+// NaN returns an IEEE 754 "quiet not-a-number" value with the
+// payload specified in the low 51 bits of payload.
+// The NaN returned by math.NaN has a bit pattern equal to NaN(1).
+func NaN(payload uint64) float64 {
```
Comment (btracey) [inline floats/floats.go]: Just wondering if this should be `NaNWith` or something to help distinguish it a bit from `NaN`

File: floats/floats.go [hunk 4]
```diff
@@ -490,6 +490,30 @@ func MulTo(dst, s, t []float64) []float64 {
...
 }
 
+const (
+	nanBits = 0x7ff8000000000000
+	nanMask = 0xfff8000000000000
+)
+
+// NaN returns an IEEE 754 "quiet not-a-number" value with the
+// payload specified in the low 51 bits of payload.
+// The NaN returned by math.NaN has a bit pattern equal to NaN(1).
```
[the remark this answers is not in the document]
Comment (kortschak) [reply floats/floats.go]: It's not guaranteed. Nor is the behaviour of `math.Copysign`.

File: floats/floats.go (hunk 3 above)
Comment (kortschak) [reply floats/floats.go]: Nice. Done.

File: floats/floats.go (hunk 1 above)
Comment (kortschak) [reply floats/floats.go]: Fear of parentheses? Done.

File: floats/floats.go (hunk 2 above)
Comment (kortschak) [reply floats/floats.go]: I think you are confusing `&` and `&^`. We want to keep all bits in `b` that are zero in `nanMask`.

**Comment by kortschak** [comment]
PTAL

File: floats/floats.go (hunk 2 above)
Comment (btracey) [reply floats/floats.go]: Ah, yes. Thanks.

**Comment by kortschak** [comment]
Sorry, just fixed a comment that became incorrect after the last commit. PTAL.

**Comment by btracey** [comment]
Oops. Sorry

**Comment by kortschak** [comment]
An alternative signature pair is `NaNWith(payload uin64, signalling bool) float64` and `NaNPayload(f float64) (payload uint64, signalling, ok bool)`. In this case the payloads would be normalised by doing a shift of `payload` by one bit if `signalling` is true. This obviates the issue of failing, but adds complexity. I think probably we just leave it as it is now.

**Comment by btracey** [comment]
I think it's better as is for now. As far as I understand, one of the main things this attempts to enable is to use these as a "not available" value. Additionally, as I understand, those kinds of values should also use quiet NaNs. Seems best to leave simple until we have another reason.

File: floats/floats.go [hunk 5]
```diff
@@ -490,6 +490,30 @@ func MulTo(dst, s, t []float64) []float64 {
...
+func NaN(payload uint64) float64 {
+	payload &^= nanMask
+	return math.Float64frombits(nanBits | payload)
+}
+
+// NaNPayload returns the lowest 51 bits of an IEEE 754 "quiet
+// not-a-number" and true, or zero and false if f is not NaN
+// or is not quiet.
```
[the remark this answers is not in the document]
Comment (kortschak) [reply floats/floats.go]: The size of the payload is important and the phrasing "X returns ...." is conventional. How about:

```
// NaNPayload returns the lowest 51 bits payload of an IEEE 754 "quiet not-a-number".
// For values of f other than quiet-NaN, NaNPayload returns zero and false.
```

[the remark this answers is not in the document]
Comment (Kunde21) [reply floats/floats.go]: That reads much better.  endorsed

[the remark this answers is not in the document]
Comment (kortschak) [reply floats/floats.go]: Sent. Please reapprove.

**Pull request merged**